### PR TITLE
Muon Analysis - Better validation of start and end x

### LINKED ISSUE #31505: Muon Analysis - Limited validation of the start and end X line edits
### Description
In Muon Analysis and FDA there is an issue with not enough validation of the start and end x.

From Steve:
```
Another issue I’ve come across is connected with the boxes used to define the fit range in the Analysis interface.
 
1.       It seems to be possible to specify a ‘time end’ beyond the end of the data. What happens when this is done?
2.       It seems to be possible to leave ‘time start’ blank, which causes Mantid to bring up the crash window
3.       It seem to be possible to leave ‘time end’ blank … the box stays blank, what happens?
4.       Whether a value as been entered and accepted seems a little indetermined.
```

Also the error reporter shows:
```
 Traceback (most recent call last):
 File "C:\MantidNightlyInstall\scripts\Muon\GUI\Common\fitting_widgets\basic_fitting\basic_fitting_presenter.py", line 219, in handle_end_x_updated
   if self.view.end_x < self.view.start_x:
 File "C:\MantidNightlyInstall\scripts\Muon\GUI\Common\fitting_widgets\basic_fitting\basic_fitting_view.py", line 126, in start_x
   return self.fit_function_options.start_x
 File "C:\MantidNightlyInstall\scripts\Muon\GUI\Common\fitting_widgets\basic_fitting\fit_function_options_view.py", line 147, in start_x
   return float(self.start_x_line_edit.text())
ValueError: could not convert string to float: ''
```

### Platforms affected
All

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR fixes a couple of things:
- The time start and time end will not go past the x limits of the data which is loaded (0.0 to 34.0 ish for Muon Analysis)
- Entering an empty string, or broken e notation, will no longer cause a crash/error
- If a start X is provided and larger than end X, the start and end X are swapped. And vice versa if end X < start X.

The LineEditDoubleValidator now being used is already well tested and is used in other areas
https://github.com/mantidproject/mantid/blob/d0872d5641a9be426bd5d04fd556e3e61e854e97/qt/python/mantidqt/utils/qt/test/test_line_edit_double_validator.py#L19

**To test:**
1. Muon Analysis
2. MUSR 62260
3. Go to fitting tab.
4. Try make time start smaller than time end. The values should swap
4. Try make time end smaller than time start. The values should swap
5. Try enter an empty string or broken e notation in the time start and end. The previous value should be reused
6. Try enter a value for start and end X which is outside the x range of the data. (The data x range is usually between 0 and 34)

Fixes #31505

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py [hunk 1]
```diff
@@ -409,6 +409,15 @@ def do_rebin(self) -> bool:
         """Returns true if rebin is selected within the context."""
         return self.context._do_rebin()
 
+    def x_limits_of_current_dataset(self) -> tuple:
+        """Returns the x data limits of the currently selected dataset."""
+        if self.current_dataset_index is not None:
+            workspace = AnalysisDataService.retrieve(self.current_dataset_name)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:415]: use this instead https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/Common/ADSHandler/ADS_calls.py#L19

then we can keep all of the ADS calls to 1 file

**Comment by AnthonyLim23** [comment]
do we have a test for max <= min?

**Comment by AnthonyLim23** [comment]
I can set both start and end x to the same value (in this case 5), but when I click fit it uses the previous value. However, its not obvious from the GUI that is what it will do

**Comment by AnthonyLim23** [comment]
In the case of the user value is less/greater than the possible values, should we set it to the min/max? I would assume that they would want to fit the whole range and just used a number thats overkill.

**Comment by robertapplin** [comment]
> I can set both start and end x to the same value (in this case 5), but when I click fit it uses the previous value. However, its not obvious from the GUI that is what it will do

It would be easier to fix this by allowing start and end x to be equal and having this fact reflected in the model, do you think that would be ok?

**Comment by AnthonyLim23** [comment]
loading PSI data causes lots of strange things. 

- The start value for the fit is 15, but that data ends at about 10. 
- If I change the end value to 10, it complains and resets to 15

**Comment by AnthonyLim23** [comment]
> > I can set both start and end x to the same value (in this case 5), but when I click fit it uses the previous value. However, its not obvious from the GUI that is what it will do
> 
> It would be easier to fix this by allowing start and end x to be equal and having this fact reflected in the model, do you think that would be ok?

how do I fit to zero data points?

**Comment by robertapplin** [comment]
> > > I can set both start and end x to the same value (in this case 5), but when I click fit it uses the previous value. However, its not obvious from the GUI that is what it will do
> > 
> > 
> > It would be easier to fix this by allowing start and end x to be equal and having this fact reflected in the model, do you think that would be ok?
> 
> how do I fit to zero data points? We could just colour them both in red and disable the fit button - to prompt the user to change it. Could even add a message somewhere

Just found an easy way to do it so its fine

**Comment by robertapplin** [comment]
Its ready for another test @AnthonyLim23

File: scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py [hunk 2]
```diff
@@ -491,7 +491,8 @@ def _get_new_end_x_for(self, new_dataset_name: str) -> float:
         if new_dataset_name in self.dataset_names:
             return self.end_xs[self.dataset_names.index(new_dataset_name)]
         else:
-            return self.current_end_x
+            x_lower, x_upper = self.x_limits_of_workspace(new_dataset_name)
+            return self.current_end_x if x_lower < self.current_end_x < x_upper else x_upper
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:495]: I dont think we can do 2 comparisons in this line ` x_lower < self.current_end_x < x_upper` unless I have missed something

Comment (robertapplin) [reply scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:495]: I think this should be fine, I have tested it and it doesn't complain. When I do the comparisons like `self.current_end_x > x_lower and self.current_end_x < x_upper` pycharm tells me to write it the above way

Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/fitting_widgets/basic_fitting/basic_fitting_model.py:495]: huh, must be a newish thing

**Review by AnthonyLim23 — APPROVED** [review]
It works well and I think its a nice usability improvement. It also stops users getting themselves into strange situations. :shipit:

**Pull request merged**